### PR TITLE
Support precomputed reference log-probs with fused DPO

### LINKED ISSUE #7083: Support precomputed reference log-probs in fused DPO training
### Problem

`DPOConfig(precompute_ref_log_probs=True, use_liger_kernel=True)` is rejected even though cached reference scores can be consumed directly by the fused preference loss. This forces every DPO training step to run a second reference-model backbone and LM-head projection.

The fused DPO path also calls the unwrapped backbone outside the mixed-precision wrapper, so BF16/FP16 configurations can silently execute that path in FP32. Finally, constructing the non-shuffling precompute DataLoader currently advances global RNG state and changes the later training sample order.

### Proposed behavior

- Let the fused DPO loss accept paired cached chosen/rejected reference log-probs and skip the reference forward.
- Restore the Accelerate autocast context around the fused path.
- Give the precompute DataLoader a private seeded generator so precomputation does not perturb training order.
- Cover cached-reference learning, autocast, and RNG preservation with tests.

Vision DPO remains outside this request because dynamic multimodal preprocessing is not currently precomputed by DPOTrainer.

### PR DESCRIPTION
**Pull request by alay2shah**
# What does this PR do?

Allows `DPOTrainer` to combine `precompute_ref_log_probs=True` with `use_liger_kernel=True`.

The fused path now forwards cached chosen/rejected reference log-probabilities into the vendored DPO loss and skips the reference backbone and LM-head projection during training. It also restores the Accelerate autocast context around the unwrapped fused path; without it, mixed-precision configurations could execute the backbone and loss in FP32.

For cached references, the vendored loss computes only target-token log-probabilities instead of materializing the full FP32 log-softmax tensor, detaches logging-only mean logits early, and skips unused NLL work. It uses the Liger selective-log-softmax kernel when available and a compatible row-wise fallback for older Liger releases and CPU. Small BF16/FP16 sigmoid workloads use the lower-overhead native-autograd route; the dispatch budget counts both policy and reference projections when references are online. Larger and non-sigmoid online-reference workloads keep the established compiled/chunked path. Large cached-reference and ZeRO-3 workloads use pair-chunked, storage-backed autograd so the selective Triton kernel is not nested under `torch.func`.

Under DeepSpeed ZeRO-3, cached-reference loss computation uses the established forward-computed custom-autograd route while the directly accessed LM head is gathered. Cached scores are split on the same chosen/rejected pair boundaries as policy inputs, and storage-backed `torch.autograd.grad` preserves selective log-softmax without the unsupported Triton-under-functorch composition. A real 2×H100 ZeRO-3 + FA2 run completed three optimizer steps with return code 0 and identical finite loss and gradient histories on both ranks.

Reference precomputation now uses a private seeded `torch.Generator`. Creating its non-shuffling DataLoader therefore no longer advances global RNG state and changes the subsequent training sample order. The same RNG fix is applied to KTO, whose duplicated precompute loader had the same behavior.

Liger DPO rejects `ld_alpha` explicitly because the fused loss does not implement LD-DPO token weighting; accepting it would combine LD-weighted cached reference scores with unweighted policy scores. Vision DPO remains unchanged because its dynamic multimodal preprocessing is not precomputed by `DPOTrainer`.

Tests cover an actual cached-reference training update, CUDA autocast, DPO and KTO RNG preservation, LD-DPO incompatibility, projection-aware dispatch, explicit native-versus-ZeRO-3 loss/metric/gradient parity, and the optimized loss route. Five focused H100 trainer tests pass, the corrected CPU native-versus-ZeRO-3 parity tests pass, and the real 2×H100 ZeRO-3 smoke passed with FA2.

Fixes #7083

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that is the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that is the case.
- [x] Did you make sure to update the documentation with your changes? (No documentation change is needed; both public flags already exist.)
- [x] Did you write any new necessary tests?

## AI writing disclosure

- [ ] No AI usage
- [ ] AI-assisted
- [x] AI-generated

---

> [!NOTE]
> **Medium Risk**
> Touches core DPO training, fused loss autograd dispatch, and DeepSpeed ZeRO-3 LM-head gathering; incorrect routing could change loss, gradients, or precision behavior for Liger + precompute setups.
> 
> **Overview**
> **Enables `DPOTrainer` to use `precompute_ref_log_probs=True` together with `use_liger_kernel=True`**, forwarding cached `ref_chosen_logps` / `ref_rejected_logps` into the fused loss and skipping the reference backbone during training when refs are precomputed.
> 
> The vendored **fused DPO loss** gains routing between a small-shape **native autograd** path (BF16/FP16, projection-size budget), a **cached-reference** path with storage-backed autograd for ZeRO-3 and large workloads, and the existing chunked Triton path. Cached refs use **selective log-softmax** (Liger kernel when present) instead of full-vocab log-softmax, with validation that precomputed refs are not mixed with live ref weights.
> 
> **`DPOTrainer`** wraps the Liger loss path in **`accelerator.autocast()`** so mixed-precision training does not silently run in FP32, rejects **`ld_alpha`** with Liger (fused loss has no LD-DPO weighting), and seeds the reference precompute **DataLoader** so it no longer advances global RNG. The same RNG fix is applied in **KTO**.
> 
> Tests cover Liger + precomputed refs training, CUDA autocast, RNG preservation, `ld_alpha` incompatibility, dispatch heuristics, and ZeRO-3 vs native parity for cached-reference loss/gradients.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit ce1dc85a0444f4679e0fd40ff9934a03fbecc080. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/trainer/dpo_trainer.py [hunk 1]
```diff
@@ -1308,7 +1309,15 @@ def _compute_loss_liger(self, model, inputs, return_outputs):
...
-                weight, hidden_states, labels, bias, ref_hidden_states, ref_weight, ref_bias
+                weight,
+                hidden_states,
+                labels,
+                bias,
+                ref_hidden_states,
+                ref_weight,
+                ref_bias,
+                ref_chosen_logps=inputs.get("ref_chosen_logps"),
+                ref_rejected_logps=inputs.get("ref_rejected_logps"),
```
[the remark this answers is not in the document]
Comment (alay2shah) [reply trl/trainer/dpo_trainer.py:1325]: Fixed in b4512187. The trainer now rejects use_liger_kernel with any non-None ld_alpha before training because the fused loss does not implement LD-DPO token weighting. A regression test covers the incompatibility. The expanded H100 suite passes.

File: trl/trainer/dpo_trainer.py [hunk 2]
```diff
@@ -1140,13 +1135,15 @@ def _precompute_ref_logps(self, dataset: Dataset, name: str, batch_size: int) ->
...
 
+        data_seed = self.args.data_seed if self.args.data_seed is not None else self.args.seed
         dataloader = DataLoader(
             dataset,
             batch_size=batch_size,
             collate_fn=self.data_collator,
             num_workers=self.args.dataloader_num_workers,
             pin_memory=self.args.dataloader_pin_memory,
             shuffle=False,
+            generator=torch.Generator().manual_seed(data_seed),
```
[the remark this answers is not in the document]
Comment (alay2shah) [reply trl/trainer/dpo_trainer.py:1151]: Fixed in b4512187. KTO now gives its precompute DataLoader the same private data-seeded generator, with a matching RNG-preservation regression test. The combined DPO/KTO H100 suite passes 5 focused tests.

File: trl/losses/dpo_loss.py [hunk 3]
```diff
@@ -333,7 +333,44 @@ def forward(
...
+                average_log_prob=self.average_log_prob,
+                loss_type=self.loss_type,
+                label_smoothing=self.label_smoothing,
+                discopop_tau=self.discopop_tau,
+                ref_chosen_logps=ref_chosen_logps,
+                ref_rejected_logps=ref_rejected_logps,
+            )
+            # Mean logits are logging-only; do not retain their graph after returning from the loss.
+            outputs = (outputs[0], outputs[1], outputs[2].detach(), outputs[3].detach(), *outputs[4:])
+            return loss, outputs
```
[the remark this answers is not in the document]
Comment (alay2shah) [reply trl/losses/dpo_loss.py:468]: Fixed in add131d0. Cached-reference DPO now detects a ZeRO-3-managed LM-head parameter and uses a forward-computed custom-autograd function, so gradients are computed and stashed while `maybe_gather_lm_head_ctx` still holds the full parameter. Native autograd remains the lower-overhead path for non-ZeRO execution. Added loss/metric/gradient parity coverage and validated an actual 2×H100 ZeRO-3 cached-reference training run for three optimizer steps with identical finite histories on both ranks.

File: trl/losses/dpo_loss.py [hunk 4]
```diff
@@ -26,6 +26,16 @@
...
 
 
+_NATIVE_DPO_MAX_LOGIT_ELEMENTS = 268_435_456
+
+
+def _should_use_native_dpo(_input, weight):
+    if _input.dtype not in (torch.float16, torch.bfloat16):
+        return False
+    total_logit_elements = (_input.numel() // _input.shape[-1]) * weight.shape[0]
+    return total_logit_elements <= _NATIVE_DPO_MAX_LOGIT_ELEMENTS
```
[the remark this answers is not in the document]
Comment (alay2shah) [reply trl/losses/dpo_loss.py:37]: Fixed in add131d0. The heuristic now multiplies logical logit elements by two when an online reference projection is present; cached references count only the policy projection. The exact 268,435,456 boundary is covered by a meta-tensor unit test.

File: trl/losses/fused_linear_preference.py [hunk 5]
```diff
@@ -27,6 +27,24 @@
...
 
+def _selective_log_softmax(logits, target):
+    if logits.device.type == "cuda":
+        try:
+            # Liger is optional, so keep this import on the Liger-only path.
+            from liger_kernel.ops.selective_log_softmax import liger_selective_log_softmax
+        except ImportError:
+            pass
+        else:
+            return liger_selective_log_softmax(logits, target)
```
[the remark this answers is not in the document]
Comment (alay2shah) [reply trl/losses/fused_linear_preference.py:38]: This is already FP32 internally. `liger_selective_log_softmax` loads BF16/FP16 logits and explicitly converts the selected values to `tl.float32` before max/exp/sum in the Triton forward kernel; its backward also operates on FP32 probabilities. The wrapper intentionally avoids materializing a full FP32 `[B,T,V]` tensor while preserving FP32 softmax arithmetic. The CPU fallback likewise calls `F.log_softmax(row_logits.float(), ...)`.

File: trl/losses/dpo_loss.py [hunk 6]
```diff
@@ -253,6 +269,50 @@ def backward(ctx, *grad_output):
...
+            compiled=compiled,
+            use_ref_model=False,
+            average_log_prob=average_log_prob,
+            chunk_size=max(1, target.shape[0] // 2),
+            loss_type=loss_type,
+            label_smoothing=label_smoothing,
+            discopop_tau=discopop_tau,
+            ref_chosen_logps=ref_chosen_logps,
+            ref_rejected_logps=ref_rejected_logps,
+        )
```
[the remark this answers is not in the document]
Comment (alay2shah) [reply trl/losses/dpo_loss.py:313]: Fixed in e395e564. Cached refs are split on the exact chosen/rejected pair chunk boundaries, `chunk_size` is preserved, and target-only selective log-softmax remains enabled. The selective branch now uses storage-backed `torch.autograd.grad`, so it stays memory-bounded without nesting the Triton autograd function under `torch.func`.

File: trl/losses/dpo_loss.py [hunk 7]
```diff
@@ -253,6 +269,52 @@ def backward(ctx, *grad_output):
...
+            bias=bias,
+            ignore_index=ignore_index,
+            beta=beta,
+            alpha=alpha,
+            compute_nll_loss=compute_nll_loss,
+            compiled=compiled,
+            use_ref_model=False,
+            average_log_prob=average_log_prob,
+            chunk_size=chunk_size,
+            selective_log_softmax=True,
```
[the remark this answers is not in the document]
Comment (alay2shah) [reply trl/losses/dpo_loss.py:306]: Fixed in e395e564. The selective cached path no longer invokes the Triton autograd function under `torch.func.grad_and_value`; it computes the per-chunk gradients with ordinary storage-backed `torch.autograd.grad` inside the outer forward-computed autograd function. A real 2xH100 ZeRO-3 + FA2 cached-reference run completed 3 optimizer steps with return code 0 and identical finite loss/gradient histories on both ranks.

File: tests/test_dpo_loss.py [hunk 8]
```diff
@@ -1445,3 +1446,46 @@ def test_label_smoothing_validation():
...
+    ref_chosen_logps = torch.randn(2, device=device)
+    ref_rejected_logps = torch.randn(2, device=device)
+
+    def run(zero3_parameter):
+        _input = base_input.detach().clone().requires_grad_(True)
+        weight = base_weight.detach().clone().requires_grad_(True)
+        bias = base_bias.detach().clone().requires_grad_(True)
+        if zero3_parameter:
+            weight.ds_id = 0
+        loss, outputs = FusedLinearDPOLoss()(
```
[the remark this answers is not in the document]
Comment (alay2shah) [reply tests/test_dpo_loss.py:1475]: Fixed in ce1dc85a. The test now explicitly forces `_should_use_native_dpo` true: the ordinary cached run therefore exercises native autograd, while the `ds_id`-marked run still takes the ZeRO-3 storage-backed custom-autograd route. The corrected focused suite passes.

**Comment by qgallouedec** [comment]
Closing: we don't review fully AI-generated PRs from first-time contributors, see the [AI usage policy](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#ai-usage-policy).
